File: python/FindNumberOfLIS.py
class Solution(object):
    def findNumberOfLIS(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """
        len_array = len(nums)
        if len_array == 0:
            return 0
        length = [0] * len_array
        count = [0] * len_array
        
        for i in range(len_array):
            new_count = 1
            new_length = 1
            for j in range(i):
                if nums[j] < nums[i]:
                    if length[j] + 1 > new_length:
                        new_length = length[j] + 1
                        new_count = count[j]
                    elif length[j] + 1 == new_length:
                        new_count += count[j]
            
            length[i] =  new_length
            count[i] = new_count
        max_len = max(length)
        max_count = 0
        for i in range(len_array):
            if length[i] == max_len:
                max_count += count[i]
        return max_count

File: c++/FindNumberOfLIS.cpp
class Solution
{
  public:
    int findNumberOfLIS(vector<int> &nums)
    {
        int n = nums.size();
        if (n == 0)
        {
            return 0;
        }
        int d[n][2];
        int ans = 0;
        int ma = 0;
        for (int i = 0; i < n; i++)
        {
            d[i][0] = 1; // i,0 is max length
            d[i][1] = 1; // i,1 is count
            for (int j = 0; j < i; j++)
            {
                if (nums[i] > nums[j])
                {
                    if (d[j][0] + 1 > d[i][0])
                    {
                        d[i][0] = d[j][0] + 1;
                        d[i][1] = d[j][1];
                    }
                    else if (d[j][0] + 1 == d[i][0])
                    {
                        d[i][1] += d[j][1];
                    }
                }
            }
            if (d[i][0] > ma)
            {
                ma = d[i][0];
            }
        }

        for (int i = 0; i < n; i++)
        {
            if (d[i][0] == ma)
            {
                ans += d[i][1];
            }
        }
        return ans;
    }
};